controllers/capabilities/openapi_controller.go

```diff
@@ -358,6 +364,55 @@ func (r *OpenAPIReconciler) validateOpenAPIAs3scaleProduct(openapiCR *capabiliti
 	return nil
 }
 
+func (r *OpenAPIReconciler) validateOASExtensions(openapiObj *openapi3.T) error {
+	extensionErrors := field.ErrorList{}
+	productExtensionPath := field.NewPath("x-3scale-product")
+	metricsExtensionPath := productExtensionPath.Child("metrics")
+	policiesExtensionPath := productExtensionPath.Child("policies")
+
+	// Validate OAS root product extension
+	rootProductExtension, err := helper.NewOasRootProductExtension(openapiObj)
+	if err != nil {
+		return err
+	}
+
+	// Validate metrics
+	if rootProductExtension != nil && rootProductExtension.Metrics != nil {
+		// Loop through policies in extension and create PolicyConfig objects
+		for metricKey, metric := range rootProductExtension.Metrics {
```

Comment:
just something to think about, not sure whether this is valid but I can create multiple metrics in my OAS with same metric ID:
```
    metric01:
      friendlyName: "My Metric 01"
      unit: "hits"
      description: "This is a custom metric"
    metric01:
      friendlyName: "My Metric 02"
      unit: "hits"
      description: "This is a custom metric 2"
```
This results in a single metric - metric01, with details from metric 2. 
I think we could check if metric if provided system name already exists, and if so, error out on OpenAPI CR.

Reply:
possibly same scenario with other extensions.
Maybe a good approach here would be to filter out the "must be unique" values and error out if uniqueness is not maintained.

Reply:
Thanks for catching this. The reason why you're getting a single metric is Go won't let you read duplicate keys into a map so when it parses the OAS from json/yaml it silently drops one. I'm not sure the best way to handle this but I agree we need to address it. Lets discuss offline

Reply:
This is actually a limitation with OAS's Go library. When [constructing](https://github.com/3scale/3scale-operator/blob/573623e1976f87db295ebc84b19688684945f3de/controllers/capabilities/openapi_controller.go#L310) the openapi object from yaml, the OAS spec is first converted to json and then unmarshaled into the openapi object. When this happens any duplicate keys are overridden and there's unfortunately no way to modify this behavior. 

So it looks like we're going to have to rely on the [documentation](https://github.com/carlkyrillos/3scale-operator/blob/5327f0ffb2428679958c4f565006075bd7ccde8d/doc/openapi-3scale-extensions.md#L16) to prevent users from creating metrics (or methods/applicationplans) with duplicate system names.